### ISSUE TITLE
General Query on Civetweb on Zephyr OS

### ISSUE THREAD

**Issue by prayassamriya** [civetweb/civetweb#1022]
Hello All, I am exploring Civetweb to get HTTP/S Webserver functionality on stm32 mcu having ( 256 KB+ RAM and 1 MB+ Flash).Can you please provide me answers to my query, It will help me decide whether I should consider it in my commercial project

1. What is the minimum ( and typical if possible ) ROM and Ram requirement to have HTTP Webserver running with max 5 socket connections to support?
2. Does Civetweb uses HEAP or static memory?
3. Is MbedTLS + Civetweb support available to get HTTPS communication?
4. Can both HTTP and HTTPS run together?
5. I have read somewhere that each socket needs a separate OS thread that means lots of RAM requirement. Is there a configuration to support just 1( master ) or 2( master + worker ) thread and still use 5 or more socket connections.
5. How does it access the webpages? If I store webpages( big html and js files ) on external flash, will it be able to support the design?
6. Does it support auto redirection to HTTPS if HTTP is disabled
7. Is it possible to enable and disable webserver at run time?
8. Does it support cybersecurity features like IP whitelisting, API limiting, audit logging( login, logout )?
9. What different authentication methods server ca support? Just Digest authentication?
10. If the complete web UI load includes multiple files ( few big, most other small in size ) , can it support time slicing where it serve and download multiple files in parallel on multiple sockets?
11. Is all of above supported on HTTP and Websocket both?

**Comment by bel2125** [comment]
I am not an expert on Zephyr - in fact I never used it - so I can only give you answers from the CivetWeb point of view.

1) Very different for HTTP and HTTPS, since you need an additional TLS library for HTTPS.  For CivetWeb itself it depends on the features you want to add - there are several feature defines you can set or unset when you compile civetweb. The default is some intermediate setting: e.g., include HTTPS and CGI but exclude IPv6 and WebSockets. If you consider this default settings "typical", a x64 Linux Binary will have around 175kB file size. You can disable CGI, SSL, ... and you can even disable serving static files to have a binary of less than 140kB ... you can go further down to disable the file system access at all ...
This depends very much on the features you need.
Of course file size does not tell you much about the total memory use at runtime ... but this depends even more on configurations.

2) Heap, stack and static memory.
If you really need to, you can overload the allocation functions (malloc/free).
It uses one thread for each open connection, and the stack will contain the input/output buffer and the state machine for this connection.
The heap consumption and the number of threads will strongly depend on your configuration - therefore I cannot give you good answer for the RAM consumption in question 1 without knowing of lots of details - you can use 10 threads or 10000 threads (factor 1000). By increasing or reducing other buffers, you might gain or loose another factor of 5 or maybe 10. The default settings are rather generous (let's say: "they are good for an older desktop PC" ... to avoid support requests for "buffer too small" an "out of memory" problems with PCs younger than 10 years). 
For the use of in a small embedded system, you have to take a closer look into the configuration options and don't use the default everywhere. Embedded systems are usually built by experienced engineers who can understand the configuration options, the defaults are meant for someone who just wants to install and run a web server on Windows or Linux (for personal use) without having to understand advanced configurations.

3) Yes, mbedtls can be used as an alternative to OpenSSL. All "standard" features work with mbedtls. SNI (required for HTTP/2) is not yet implemented for mbedtls - I am sure it can be done, but it has not been done yet.

4) Yes.
You can configure HTTP as a forward to the HTTPS page or you can use both at the same time. You can also restrict HTTP to one network interface card (internal network) and use HTTPS for all networks (internal and external) .. just a matter of configuration-

5) CivetWeb uses one worker thread for each active connection. You can restrict the stack size for each worker thread to maybe 32 kB (depending on your configurations ... somewhere I already wrote some considerations about the stack size). Probably you can use less if you only require basic features and don't use a lot of stack in your own callbacks. Here it also depends on what the operating system does - is every stack page pre-commited to physical memory or is it just virtual address space. E.g., for Linux and Windows it is just virtual address space that you reserve, and you can easily reserve 1 MB of address space and still use only 20 kB of physical memory. For embedded operating systems that precommit the full stack, you have to be more carefully.
You can use only one master and one worker thread - I would recommend not to do this for performance reasons (due to the way browsers tend to access web servers ... long story, but can be tested easily). 

6) Not sure if I understand the question. To serve content, you basically have three options:
a) Put the files somewhere in some directory and configure the directory as "document_root". CivetWeb will use fopen/fread/fstat/... to access the files.
b) Write request handler (callbacks in C programming language) for ever URL you want to support. You can use a callback for a directory name to serve all "files" below. In contrast to files you can generate the content dynamically on the fly.
c) Use a script to generate the server response. This can be CGI (attention: CGI will spawn a new process for every request), or the built-in Lua interpreter. You can put the script file into the document_root folder.
d) Use a combination of a+b+c. I think it is the typical case to combine a+b or a+c and sometimes even a+b+c.

7) Yes. Set the configuration "listening_ports" to "80r,553s" - this means Port 80=redirect, Port 443=secure

8) I don't know your exact requirements and your exact environment. You can start and stop the webserver (mg_start, mg_stop API). If it is a separate executable, you can simply start and stop this executable. From my experience, often people don't want to enable/disable the entire web server, but only certain pages - this can be done using 6b and 6c above. Usually I don't register/unregister handlers, but return a 400-something HTTP error code if a certain URL is not available right now for some reasons. This all depends on your exact requirements.

9) "cybersecurity features" is a rather general term - I can only comment your examples:
"IP whitelisting": If it is a short list of allowed client IPs there is a way to configure this
"API limiting": If you mean throttling of the download rate for certain IP addresses: yes
"audit logging": You can enable an "access_log", and if you enable statistics and write some hooks/callbacks, you can do more auditing. Login/logout is usually done using cookies - this needs to be done in a callback or script (6b, 6c) that you need to provide. The CivetWeb built-in mechanism would be HTTP digest authentication - it works but many web designers don't like it, because the design of the login page is a browser standard dialog that cannot be designed and will probably look different to the (corporate-)design of the web page.
For the best possible security, you can activate client side certificates. Then the client (browser or other program) will be forced to use HTTPS *AND* will have to provide an additional certificate that has to be known by the server. Of course, certificate Management (how to upload the required certificates ... and all organizational measures) are beyond the scope of CivetWeb.

10) Details see above
HTTP Digest is built in
Cookie based auth has to be provided by a script (CGI, Lua) or a callback
HTTPS client certificates if you activate this configuration
If you use callbacks, you can also implement your own, completely unique idea how authentication and authorization checks should work. There is an API to register "auth handlers".

11) Usually you do this by using multiple worker threads. If the client is a browser, it will open multiple HTTP/1.1 connections to the same page and distribute these load calls between the different connections. Since every connection has one thread, this will effectively distribute the work and your page will load in parallel. Thus, you sould use more than one worker thread.

12) You need to activate websockets when you build the server - it is deactivated by default. Once activated, it will support all off the above. There are several additional configuration parameters for websockets. For example, you can set an additional "websocket_root" if you have scripts handling websocket requests, or you can set additional timeouts.

**Comment by bel2125** [comment]
And if you are missing a feature, that may be a valuable addition to CivetWeb:
Contributions are welcome (under the MIT license).

**Comment by prayassamriya** [comment]
@bel2125 many thanks. your response helped me understand it better. we do have our own custom webserver and my questions are based on what is supported there. We would like to migrate to civetweb on RTOS Env ( not linux or windows ) and doing some gap analysis before we make the decision.
if you don't mind I have few follow-up questions -

6.1. As our target platform is RTOS based microcontroller, using file system is a luxury. In our case, the content could be in external memory. When server receive the request, The content ( web page file ) has to be loaded into ram buffer( ram buffer could be of limited size so it should be done in reparative manner in chunks ) and then send to client. once first chunk is sent out same operation should be repeated until complete file is sent. I assume that its archivable without modifying civetweb lib.

9.1. Other CS features I can think of are 
      a) Locking user account ( for a defined time period ) if user enters wrong credentials for defined number of times and unlocking it when timeout expires.
b) Force change default password
c) Use of SHA instead of MD5 ( if in use ) in digest authentication
d) Control on Cross-Origin Resource Sharing (CORS) policy like allow all origins or not
e) Configurable User session Inactivity timeout
let me know what is supported and what is not. Or if you feel few of them are beyond the scope of civetweb but all hooks are available to achieve it in application space.

13. What is the format of the data received in callback - raw or structured ( parsed data ). I am looking for the usecase where a PUT or a POST URL has **json or xml** payload. Does Civetweb parse it and convert it into structured data or complete parsing has to be done in callback? 

Thanks for your guidance.

**Comment by bel2125** [comment]
6.1)
For this kind of systems, there are the defines "NO_FILES" and "NO_FILESYSTEM". Then all requests have to be handled by callbacks. In a callback, you need to generate the web content on request - of course you can buffer it somewhere, take it from a (shared) memory or whatever you can program in C. You can use chunked transfer encoding, if you don't know the response size when you start sending a response.
Some examples how to use callbacks can be found in examples/embedded_c/
This is all possible without modifying the source. The mentioned build time configuration options are explained in docs/Building.md and docs/Embedding.md. Load time configuration options are explained in docs/UserManual.md

9.1)

a) This is related to your auth method. If you are using client certificates, there is no password sent to the the server side. 
You probably don't want to use digest authentication - the passwords would be stored in a file, and there are the style/visual design topics explained abobe.
If you are using your own auth callback, it is up to you to implement this strategy there.

b) Again depending on your auth method.
If you are using your own auth callback, it is up to you to implement this strategy there.

c) I think this would violate the specification. HTTP digest auth uses an MD5 and nothing else.
Please have a look at https://en.wikipedia.org/wiki/Digest_access_authentication - in particular "Impact of MD5 security on digest authentication" but also "Disadvantages".
Digest authentication is as it is ... and you can use it or not.
From the CivetWeb implementation point of view, it would not be too much effort to implement SHA256 or anything else istead of MD5, but this does not help you if the browser uses MD5 according to the specification. 

d) Yes, this is possible by setting an appropriate configuration. Also CORS preflights are supported.

e) Again depending on your auth method.
If you are using your own auth callback, it is up to you to implement this strategy there.
 
For all followups of 9) I would recommend we first define one auth method, and then have a look at how to deal with all these points. I know about some applications of CivetWeb in different embedded products for at least two completely different (non-adjacent) markets with mandatory security certifications - so it is definitely possible to fulfill higher security levels according to different market specific standards. Do you have a particular standard in mind or what market are you targeting with your product? Maybe I already know that standard and can give some more targeted answer.
Or have you already done some thread analysis / risk modelling, and already defined some required countermeasure and are looking if this countermeasure can be implemented?
Ordering "one cybersecurity please" will not work ;-)
But with some more information, I might be able to give you some advice how this can be done - every free advice comes without any warranties, of course.

13)
A C callback will get raw data. You can use a library like cJSON (also MIT licensed) to parse JSON.
While this is a combination that is known to work successfully, CivetWeb does not ship with cJSON included - you have to download it separately (but there are code examples).
An alternative would be to use Lua instead of C. There are Lua libraries to parse JSON and XML - you do not need cJSON in this case. If you don't know Lua: It is a scripting language with a small footprint that can be used in embedded systems (www.lua.org) - depending on the web developers and the exact use cases, it might be easier to write callbacks in Lua than in C.

In general, CivetWeb takes care of the HTTP transport protocol, not about decoding/interpreting the content (HTTP body data). This keeps the footprint smaller, and it also keeps CivetWeb immune to cybersecurity threads like mobile code or the billion laughs attack. It can handle transfer encodings like compression (using zlib) or TLS, but will not interpret a document you sent. The mime type (from the HTTP header) and the body has to be handled by a callback or a script.
Then there are plenty of good libraries for handling certain mime types (like cJSON for application/json). There are hundreds of different mime types, but usually you don't use a lot of them in an (embedded) web application - adding all of them would bloat the code and introduce cybersecurity weaknesses for mime types you are not even using.

**Comment by prayassamriya** [comment]
@bel2125 Thank you for your answers. no more questions for now. :)

**Comment by NarendraSonawane1** [comment]
Hi I am trying to enable HTTPS server on zephyr RTOS. so i have http_server code in ciwetweb folder where i change options and call init_ssl   like this callbacks.init_ssl = init_ssl; but i am not able to call init_ssl function

**Comment by bel2125** [comment]
The `init_ssl` function is called automatically by CivetWeb if you have a configuration that requires HTTPS.
In particular, you need to configure secure `listening_ports` for example `443s` (note the `s`), and a `ssl_certificate` file.
Note that I am not an expert on Zephyr - in fact I never used it myself. All information I can give is tested for Linux and/or Windows. It *should* be the same for Zephyr (after all, it's the same C code), but I don't know if the Zephyr project set any specific compiler defines or made some modifications - for example, if `NO_SSL` is set then the HTTPS support has been disabled at build time.

**Comment by NarendraSonawane1** [comment]
unsigned char cert_der[] = {        0x30, 0x82, 0x04,.........................................}
unsigned char key_der[] = {        0x30, 0x82, 0x04, 0xa3, 0x02.......................................}

static int init_ssl(void *ssl_context, void *user_data){  struct ssl_ctx_st *ctx = (struct ssl_ctx_st *)ssl_context;printf("in Init ssl function....\n");return 0;
}
void *main_pthread(void *arg){        struct mg_callbacks callbacks;        struct mg_context *ctx;
        (void)arg        memset(&callbacks , 0, sizeof(callbacks));
        callbacks.init_ssl =  init_ssl;
        printf("here*********");        static const char * const options[] = {                "listening_ports",                "443s",                "ssl_certificate",                (const char *)cert_der,                "num_threads",                "1",                "max_request_size",                STRINGIFY(MAX_REQUEST_SIZE_BYTES),                "enable_auth_domain_check",                "no",                //"ssl_protocol_version",                //"TLSv1_2",                NULL        };
        ctx = mg_start(&callbacks, 0, (const char **)options);
        if (ctx == NULL) {                printf("Unable to start the server.");                return 0;        }        mg_set_request_handler(ctx, "/info$", system_info_handler, 0);       return 0;}
Thanks for support.. My code looks like this as I dont have file system on board so i passed key and cert in array format as shown above.. as ciwetweb of zephyr RTOS dont have openssl support so i am trying to integrate with some embedtls API for cert and key in init_ssl ...but function call in not going to init_ssl. and value of printf of init_ssl in not getting printedcan you please help me here    On Friday, 3 February, 2023 at 11:50:37 am IST, bel2125 ***@***.***> wrote:  
 
 


The init_ssl function is called automatically by CivetWeb if you have a configuration that requires HTTPS.
In particular, you need to configure secure listening_ports for example 443s (note the s), and a ssl_certificate file.
Note that I am not an expert on Zephyr - in fact I never used it myself. All information I can give is tested for Linux and/or Windows. It should be the same for Zephyr (after all, it's the same C code), but I don't know if the Zephyr project set any specific compiler defines or made some modifications - for example, if NO_SSL is set then the HTTPS support has been disabled at build time.

**Comment by bel2125** [comment]
Hint: You can avoid GitHub messing up the format: to embed code in a message, use three ```  `  ``` characters.

There is a full example how to embed a certificate for a system without a file system here:
https://github.com/civetweb/civetweb/blob/master/examples/embed_certificate/ec_example.c

The sample uses the OpenSSL calls `SSL_CTX_use_certificate_ASN1` call.
For mbedtls, this must be replaced by a corresponding call (not sure: `mbedtls_x509_parse` ?), but otherwise the code should be similar to the example.
